WFS/OAPIF: fix filtering on duplicated layers (fixes #48465)

rouault opened the pull request:
This is a continuation of #48500 that fixes the original issue, and passes regression tests

3nids committed:
use a QExplicitlySharedDataPointer for WFS shared data on the provider

hence we can detach when setting the subset string since it's stored in the shared data

fixes #48465

m-kuhn committed:
Create a deep copy of wfs shared data on setSubsetString

rouault committed:
[WFS provider] Fix clone of shared data, and do it only in handlePostCloneOperations() (fixes #48465)

rouault committed:
[OAPIF provider] Fix clone of shared data, and do it only in handlePostCloneOperations()

m-kuhn commented:
Thanks for the help.
Something that's still on my todo list is checking for connections between signal/slots between layer and shared data.

pathmapper commented:
Thanks to all, confirmed this fixes the issue for WFS and OAPIF 👍

rouault commented:
@m-kuhn maybe we should merge this PR as it, pending your follow ups  ?

m-kuhn commented:
This is new code for me, I am not confident to make the call here.

Given that so far we just copied the shared ptr without connecting it to the cloned layer, I wonder how important that even is (it's for some feedback loop for feature downloading dialogs etc showing up IIRC ?)

rouault committed:
[WFS/OPAIF provider] Handle reconnecting signals to shared data in handlePostCloneOperations()

rouault commented:
@m-kuhn  not completely sure if that's what you had in mind but I've added a commit to handle reconnecting the cloned shared data to the provider

m-kuhn reviewed (approved):
That is exactly what I had in mind, thank you very much

m-kuhn commented on src/providers/wfs/qgsoapifprovider.cpp:
```diff
@@ -337,7 +337,14 @@ const QString &QgsOapifProvider::clientSideFilterExpression() const
+  disconnect( mShared.get(), &QgsOapifSharedData::extentUpdated, this, &QgsOapifProvider::fullExtentCalculated );
+
+  // We must not change the subset string of the shared data used in another iterator/data provider ...
+  mShared.reset( qobject_cast<QgsOapifProvider *>( source )->mShared->clone() );
```
Can we stick to doing that just after set subset string or not?

If it's done in here we run some risks again:
- we have the subset string in a shared data that is changed while iterators are running on it, this looks risky to me
- feature ids will be changed on cloned layers which was the original issue related to the label export algorithm having wrong references to original features

can we move this only to post-subset-string change?

rouault replied:
> can we move this only to post-subset-string change?

I actually tried this. But there was one unit test that breaks in [providertestbase.py](https://github.com/qgis/QGIS/pull/48998/files#diff-e112d0d5d27c1dfff4ecd30f58bb07a19c07051a8f2c71ae4f9dbadf85146f77) at the place where I left a comment. This test is related to feature id stability. The fact that this test passes with the current code is mostly a miracle (actually when running this setSubsetString() test in isolation and not the whole test_provider_wfs.py file, it breaks, probably because it depends on other tests to have previously loaded the whole layer in the cache). So maybe we should do just a hack here and not try this test for the WFS provider

Regarding "feature ids will be changed on cloned layers", there might be an issue in using the same SharedData object by 2 layers. If the 2 layers are rendered in separate threads, the SharedData would be used (and the objects it contains, in particular the FID<->gml:id cache) concurrently, and bad things could happen. So it seems safer to me to clone the SharedData as a post processing step of layer cloning. Ideally we'd want to duplicate the FID<-->gml:id cache (which resides in a SQLite db), but that isn't not implemented currently.

m-kuhn replied:
> > can we move this only to post-subset-string change?
> 
> I actually tried this. But there was one unit test that breaks in [providertestbase.py](https://github.com/qgis/QGIS/pull/48998/files#diff-e112d0d5d27c1dfff4ecd30f58bb07a19c07051a8f2c71ae4f9dbadf85146f77) at the place where I left a comment. This test is related to feature id stability. The fact that this test passes with the current code is mostly a miracle (actually when running this setSubsetString() test in isolation and not the whole test_provider_wfs.py file, it breaks, probably because it depends on other tests to have previously loaded the whole layer in the cache). So maybe we should do just a hack here and not try this test for the WFS provider

Skipping this test sounds reasonable with this explanation.

> Regarding "feature ids will be changed on cloned layers", there might be an issue in using the same SharedData object by 2 layers. If the 2 layers are rendered in separate threads, the SharedData would be used (and the objects it contains, in particular the FID<->gml:id cache) concurrently, and bad things could happen. So it seems safer to me to clone the SharedData as a post processing step of layer cloning. Ideally we'd want to duplicate the FID<-->gml:id cache (which resides in a SQLite db), but that isn't not implemented currently.

One single layer can already now have iterators running in different threads, so concurrency shouldn't be something new to handle, unless I am missing some of the magic of the WFS provider.

rouault replied:
> can we move this only to post-subset-string change?

I've implemented that in the latest commit

> One single layer can already now have iterators running in different threads, so concurrency shouldn't be something new to handle

hum, there might be potential troubles then as I doubt the code is ready for that. (and that would be probably hard to achieve)

m-kuhn replied:
Ah, sorry I have missed this comment.
In this case I would say let's merge this, it only improves the situation.
In parallel, should we file a ticket to review multithreading here for an upcoming bugfix sprint?

rouault replied:
> In parallel, should we file a ticket to review multithreading here for an upcoming bugfix sprint?

done in https://github.com/qgis/QGIS/issues/49181

rouault committed:
[WFS/OPAIF] Only clone sharedData in setSubsetString()

- Remove calls to invalidateCache() which aren't necessary since the
  clone() operation doesn't clone cache information
- Tune providertestbase.py to avoid a failure on WFS
